packages/api/src/middlewares/integration.ts

```diff
@@ -0,0 +1,92 @@
+import { TRPCError } from "@trpc/server";
+
+import { and, eq, inArray } from "@homarr/db";
+import { integrations } from "@homarr/db/schema/sqlite";
+import type { IntegrationKind } from "@homarr/definitions";
+import { z } from "@homarr/validation";
+
+import { decryptSecret } from "../router/integration";
+import { publicProcedure } from "../trpc";
+
+// TODO: add access control in some way, most likely reusable middleware or something
+
+export const createOneIntegrationMiddleware = <TKind extends IntegrationKind>(
+  ...kinds: TKind[]
+) => {
+  return publicProcedure
+    .input(z.object({ integrationId: z.string() }))
+    .use(async ({ input, ctx, next }) => {
+      const integration = await ctx.db.query.integrations.findFirst({
+        where: and(
+          eq(integrations.id, input.integrationId),
+          inArray(integrations.kind, kinds),
+        ),
+        with: {
+          secrets: true,
+        },
+      });
+
+      if (!integration) {
+        throw new TRPCError({
+          code: "NOT_FOUND",
+          message: "Integration not found",
+        });
+      }
+
+      const { secrets, kind, ...rest } = integration;
+
+      return next({
+        ctx: {
+          integration: {
+            ...rest,
+            kind: kind as TKind,
+            decryptedSecrets: secrets.map((secret) => ({
+              ...secret,
+              value: decryptSecret(secret.value),
+            })),
+          },
+        },
+      });
+    });
+};
+
+export const createManyIntegrationMiddleware = <TKind extends IntegrationKind>(
+  ...kinds: TKind[]
+) => {
+  return publicProcedure
+    .input(z.object({ integrationIds: z.array(z.string()) }))
+    .use(async ({ ctx, input, next }) => {
+      const dbIntegrations =
+        input.integrationIds.length >= 1
+          ? await ctx.db.query.integrations.findMany({
+              where: and(
+                inArray(integrations.id, input.integrationIds),
+                inArray(integrations.kind, kinds),
+              ),
+              with: {
+                secrets: true,
+              },
+            })
+          : [];
+
+      if (dbIntegrations.length !== input.integrationIds.length) {
+        throw new TRPCError({
+          code: "NOT_FOUND",
+          message: "Integration not found",
+        });
+      }
+
+      return next({
+        ctx: {
+          integrations: dbIntegrations.map(({ secrets, kind, ...rest }) => ({
+            ...rest,
+            kind: kind as TKind,
+            decryptedSecrets: secrets.map((secret) => ({
+              ...secret,
+              value: decryptSecret(secret.value),
+            })),
```

Comment:
What do you need those for?

Reply:
Those are used directly by the integrations to access the data